### ISSUE TITLE
[Feature request]: Make The Emulator Smaller

### ISSUE THREAD

**Issue by JoLiKMC** [visualboyadvance-m/visualboyadvance-m#1015]
### Request box

The executable for old versions of _VisualBoy Advance_ weighed anywhere between 1.1 MB and 2.0 MB.
The executable for the _32-bit version_ of _VisualBoy Advance-M_ currently weighs in at a staggering **27.6 MB**.
The executable for the _64-bit version_ of _VisualBoy Advance-M_ currently weighs in at a stunning **33.3 MB**.

Please, rectify.  Thank you.

### Reason box

I see no reason as to why the new version of the emulator has to be so much bigger than the old when the old one worked just as well as the new one.

### Examples

![VBA](https://user-images.githubusercontent.com/102710555/194524483-897d1184-e92d-4784-8d3f-28edc571ecbc.png)
_A size comparison of commonly available builds of VisualBoy Advance_

**Comment by rkitover** [comment]
I don't think we could go back to the sizes of the MFC builds, but we can make a significantly smaller executable.

Currently, the builds include ffmpeg and winsparkle, and winsparkle contains yet another copy of wxWidgets within itself.

winsparkle may have been a bad idea in general.

I may be able to work on making another build target without these and with other size reducing enhancements.

I can't promise a time-frame and I doubt we'd get back to the sizes of the MFC builds.

**Comment by rkitover** [comment]
Also I wanted to ask you to explain your perspective on why this is terrible on recent hardware. This is still a tiny program by modern standards.

Somewhat related to this, I want to work on making a build that runs on XP, for the people still playing with retro hardware etc..

**Comment by ZachBacon** [comment]
The only reason the footprint is larger is due to the dependencies we use now that are statically linked, even if we did go back to the MFC version, but we did the same thing in terms of statically linking, you're still looking at a larger executable size.

**Comment by ZachBacon** [comment]
also keep in mind, for distribution reasons, we just statically linked everything in, so all we need to do is distribute one executable instead of multiple dll dependencies that vba-m now uses, which is sfml, sdl2, ffmpeg for the movie recording, wxwidgets for the toolkit we now default to, it does make it a little larger, but statically linking them to the executable does make it easier to move them around in just one executable.

**Comment by JoLiKMC** [comment]
> Also I wanted to ask you to explain your perspective on why this is terrible on recent hardware. This is still a tiny program by modern standards.
> 
> Somewhat related to this, I want to work on making a build that runs on XP, for the people still playing with retro hardware etc..
I mean… you kind of answered your own question?  But, here's my take:
If it doesn't _need_ to take up a lot of space, then it shouldn't _have_ to.

Basically, I'm an old fuddy-duddy that comes from a generation where hard drives were _optional_ and _tiny_.  I can't argue that even _this_ version of the emulator could be copy-and-pasted a literally **millions** of times before it would even come _close_ to "filling up" a modern hard drive, but I mean… does that mean it's okay to _artificially inflate_ the file size with non-essentials?

> also keep in mind, for distribution reasons, we just statically linked everything in, so all we need to do is distribute one executable instead of multiple dll dependencies that vba-m now uses, which is sfml, sdl2, ffmpeg for the movie recording, wxwidgets for the toolkit we now default to, it does make it a little larger, but statically linking them to the executable does make it easier to move them around in just one executable.

And see, I honestly didn't know that.  That, to me, actually _does_ make sense.  Older versions relied on external codecs and the like – which is perfectly fine!  It's what I've been used to for almost _two decades_, now!  But for the sake of consistency _and_ convenience, yes, it makes sense to have all that packed into the files.  Maybe not in the EXE, _itself_, as that seems less-than-ideal for troubleshooting – though not giving users access to those files also _cuts down_ on troubleshooting, I suppose.  I dunno.  I'm just one of those "it should be _this way_" sorts.

I hope my (actual) rational made sense.  The given explanation makes sense to me and I appreciate you taking the time to explain, as well as read and/or consider my suggestion.  Thank you very much!

**Comment by ZachBacon** [comment]
Now that being said, we don't need all the codecs in ffmpeg, so we are gonna try and slim down our static ffmpeg to include only the necessary ones we need, so it should make it smaller for sure. Another thing that can be done, is UPX, however we don't use that because it can also generate false positives, so in the end, if a user wants a smaller executable using https://portableapps.com/apps/utilities/portableapps.com_appcompactor and it's default settings should reduce the size of the executable even further

**Comment by rkitover** [comment]
The ffmpeg the windows builds use is from vcpkg, so it's a fairly standard build. Also, during static linking the linker removes any unused functions and code paths.

winsparkle.dll is added as a resource and is 2.5MB, should probably be compressed.

The executable compresses well, the nightly zips are only about 12MB compared to 34MB for the executable, so we could consider using upx or some other exe compressor and running builds through virustotal to make sure they don't trigger anything.

**Comment by n-a-c-h** [comment]
I ran a test with UPX just to see what the potential difference is like for 2.1.5. Even if some virus scanners have a false positive, perhaps some users want this anyway as an alternate download.

```
29,717,648 vba-m-32.exe
 9,204,880 vba-m-32-compressed.exe
 9,173,668 vba-m-32-compressed.zip
35,649,680 vba-m-64.exe
11,094,160 vba-m-64-compressed.exe
11,062,866 vba-m-64-compressed.zip
```

**Comment by Squall-Leonhart** [comment]
Mudlord actually developed a custom packer that could be ideal for the task, it would need to be made part the workflow prior to signing though.

**Comment by n-a-c-h** [comment]
What are the file sizes like with his packer?

**Comment by Squall-Leonhart** [comment]
@mudlord can you show your packer results
